blocks/search.py

```diff
@@ -0,0 +1,266 @@
+from blocks.graph import ComputationGraph
+
+floatX = config.floatX
+
+
+def construct_dict(func):
+    def wrapper(self, *inputs):
+        names, outputs = func(self, *inputs)
+        return OrderedDict(zip(names, outputs))
+    return wrapper
+
+
+class BeamSearch(object):
+    """Beam search.
+
+    Parameters
+    ----------
+    beam_size : int
+        Size of beam.
+    sequence_generator : sequence generator
+        Sequence generator brick.
+    attended : theano variable
+        Theano variable for attended in sequence generator.
+    attended_mask : theano variable
+        Theano variable for attended mask in sequence generator.
+    inputs_dict : dict
+        Dictionary of inputs {name: theano variable for input}. The
+        functions will be constructed with these inputs.
+
+    """
+    def __init__(self, beam_size, sequence_generator, attended,
+                 attended_mask, inputs_dict):
+        self.beam_size = beam_size
+        self.sequence_generator = sequence_generator
+        self.attended = attended
+        self.attended_mask = attended_mask
+        self.inputs_dict = inputs_dict
+        self.generate_names = sequence_generator.generate.outputs
+        self.init_computer = None
+        self.next_computer = None
+        self.attended_computer = None
+        self.initial_state_computer = None
+        self.state_names = (sequence_generator.state_names +
+                            sequence_generator.glimpse_names +
+                            ['outputs'])
+        self.compiled = False
+
+    def compile(self, *args, **kwargs):
+        """Compiles functions for beam search."""
+        generator = self.sequence_generator
+        attended = self.attended
+        attended_mask = self.attended_mask
+
+        self.attended_computer = function(self.inputs_dict.values(),
+                                          [attended, attended_mask],
+                                          on_unused_input='ignore')
+
+        initial_states = OrderedDict()
+        for name in self.state_names:
+            initial_states[name] = generator.initial_state(
+                name,
+                self.beam_size,
+                attended=attended)
+
+        self.initial_state_computer = function([attended, attended_mask],
+                                               initial_states.values(),
+                                               on_unused_input='ignore')
+
+        # Define inputs for next values computer
+        cur_variables = OrderedDict()
+        for name, value in initial_states.iteritems():
+            cur_value = tensor.zeros_like(value)
+            cur_value.name = name
+            cur_variables[name] = cur_value
+
+        next_generated = generator.generate(attended=attended,
+                                            attended_mask=attended_mask,
+                                            iterate=False,
+                                            n_steps=1,
```

Comment:
Fortunately you do not have to specify  `n_steps` or `batch_size` anymore after #330. That was a sort of bug that in order to use `iterate=False` one had to provide fake values for them.